x-pack/plugins/upgrade_assistant/server/lib/reindexing/worker.ts

```diff
@@ -126,14 +127,42 @@ export class ReindexWorker {
     }
   };
 
-  private refresh = async () => {
+  private updateInProgressOps = async () => {
     try {
-      this.inProgressOps = await this.reindexService.findAllByStatus(ReindexStatus.inProgress);
+      const inProgressOps = await this.reindexService.findAllByStatus(ReindexStatus.inProgress);
+      const parallelOps: ReindexSavedObject[] = [];
+      const queueOps: ReindexSavedObject[] = [];
+      for (const inProgressOp of inProgressOps) {
+        if (inProgressOp.attributes.reindexOptions?.queueSettings) {
+          queueOps.push(inProgressOp);
+        } else {
+          parallelOps.push(inProgressOp);
+        }
+      }
+
+      if (queueOps.length) {
+        const [firstInQueueOp] = queueOps.sort(
+          (a, b) =>
+            a.attributes.reindexOptions!.queueSettings!.queuedAt -
+            b.attributes.reindexOptions!.queueSettings!.queuedAt
+        );
+
+        this.log.debug(
+          `Queue detected; current length ${queueOps.length}, current item ReindexOperation(id: ${firstInQueueOp.id}, indexName: ${firstInQueueOp.attributes.indexName})`
+        );
+
+        this.inProgressOps = parallelOps.concat(firstInQueueOp);
+      } else {
+        this.inProgressOps = parallelOps;
+      }
```

Comment:
This took me some time to figure out what this was supposed to do. I think this could benefit from a comment that explains the algorithm that decides which reindex operations should be processed.

Alternatively, this could benefit from extracting the logic into a named function that explains what it does:

```ts
const inProgressOps = await this.reindexService.findAllByStatus(ReindexStatus.inProgress);
const firstOpInQueue = getFirstInQueue(inProgressOps);
const unqueuedOps = inProgressOp
  .filter(op => !!inProgressOp.attributes.reindexOptions?.queueSettings);
this.inProgressOps = [...unqueuedOps, ...([firstOpInQueue] ? firstOpInQueue : [])];
```

Reply:
Thanks for taking a look @joshdover ! I'm happy the implementation (on the whole) makes sense.

I'll clarify the context in which this code runs with a comment because it's not clear from just look at it here that it runs on a schedule and what exactly "parallel" means.